### PR TITLE
A whole bunch of things to bring auth (compat) to parity

### PR DESCRIPTION
**Pull request by samhorlbeck**

### CODE REVIEW

File: packages-exp/auth-compat-exp/src/user_credential.ts [hunk 1]
```diff
@@ -116,10 +116,10 @@ export async function convertConfirmationResult(
...
   confirmationResultPromise: Promise<externs.ConfirmationResult>
 ): Promise<compat.ConfirmationResult> {
-  const { verificationId, confirm } = await confirmationResultPromise;
+  const confirmationResultExp = await confirmationResultPromise;
   return {
-    verificationId,
+    verificationId: confirmationResultExp.verificationId,
     confirm: (verificationCode: string) =>
-      convertCredential(auth, confirm(verificationCode))
+      convertCredential(auth, confirmationResultExp.confirm(verificationCode))
```
Comment (samhorlbeck) [inline packages-exp/auth-compat-exp/src/user_credential.ts:123]: just calling confirm(verificationCode) meant that `this` wasn't bound

File: packages-exp/auth-exp/src/core/auth/auth_impl.ts [hunk 2]
```diff
@@ -333,8 +387,11 @@ export class AuthImpl implements Auth, _FirebaseService {
   }
 
   async _redirectUserForId(id: string): Promise<User | null> {
-    // Make sure we've cleared any pending ppersistence actions
-    await this.queue(async () => {});
+    // Make sure we've cleared any pending ppersistence actions if we're not in
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/auth/auth_impl.ts:401]: typo: persistence

File: packages-exp/auth-exp/src/core/user/additional_user_info.ts [hunk 3]
```diff
@@ -42,10 +42,11 @@ export function _fromIdTokenResponse(
     ];
     if (signInProvider) {
       const filteredProviderId =
-        providerId !== externs.ProviderId.ANONYMOUS &&
-        providerId !== externs.ProviderId.CUSTOM
+        signInProvider !== externs.ProviderId.ANONYMOUS &&
+        signInProvider !== externs.ProviderId.CUSTOM
           ? (signInProvider as externs.ProviderId)
           : null;
+      console.log('In first branch');
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/user/additional_user_info.ts]: remove these logs please

File: packages-exp/auth-exp/src/core/user/additional_user_info.ts [hunk 4]
```diff
@@ -68,8 +69,10 @@ export function _fromIdTokenResponse(
       );
     case externs.ProviderId.CUSTOM:
     case externs.ProviderId.ANONYMOUS:
+      console.log('In second branch');
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/user/additional_user_info.ts]: also here and two lines lower

File: packages-exp/auth-exp/src/core/user/additional_user_info.ts [hunk 5]
```diff
@@ -79,7 +82,9 @@ class GenericAdditionalUserInfo implements externs.AdditionalUserInfo {
     readonly isNewUser: boolean,
     readonly providerId: externs.ProviderId | null,
     readonly profile: externs.UserProfile = {}
-  ) {}
+  ) {
+    console.log('In constructor');
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/user/additional_user_info.ts]: also here

File: packages-exp/auth-compat-exp/src/user_credential.ts (hunk 1 above)
Comment (avolkovi) [reply packages-exp/auth-compat-exp/src/user_credential.ts:123]: weird

File: packages-exp/auth-exp/src/core/user/additional_user_info.ts [hunk 6]
```diff
@@ -129,7 +134,21 @@ class TwitterAdditionalUserInfo extends FederatedAdditionalUserInfoWithUsername
...
+    return {
+      providerId: null,
+      isNewUser: false,
+      profile: null
+    };
+  }
+
+
+  const result = _fromIdTokenResponse(_tokenResponse);
+  if (result?.providerId === 'anonymous') {
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/user/additional_user_info.ts]: should this be in the enum?

File: packages-exp/auth-exp/src/core/user/reload.ts [hunk 7]
```diff
@@ -46,6 +46,13 @@ export async function _reloadWithoutSaving(user: User): Promise<void> {
   const newProviderData = coreAccount.providerUserInfo?.length
     ? extractProviderData(coreAccount.providerUserInfo)
     : [];
+
+  const providerData = mergeProviderData(user.providerData, newProviderData);
+
+  const oldIsAnonymous = user.isAnonymous;
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/user/reload.ts:52]: can you add some comments please? why do we need all this logic to figure out if the user is anonymous?

File: packages-exp/auth-exp/src/core/user/token_manager.ts [hunk 8]
```diff
@@ -56,14 +56,13 @@ export class StsTokenManager {
       return this.accessToken;
     }
 
-    if (!this.refreshToken) {
-      assert(!this.accessToken, AuthErrorCode.TOKEN_EXPIRED, {
+    if (this.accessToken && !this.refreshToken) {
+      fail(AuthErrorCode.TOKEN_EXPIRED, {
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/user/token_manager.ts:60]: can you just make the if above an assert?

File: packages-exp/auth-exp/src/model/popup_redirect.ts [hunk 9]
```diff
@@ -92,4 +92,7 @@ export interface PopupRedirectResolver extends externs.PopupRedirectResolver {
     cb: (support: boolean) => unknown
   ): void;
   _redirectPersistence: externs.Persistence;
+
+  // This is needed so that auth does not have a hard dependency on redirect
+  _completeRedirectFn: (auth: externs.Auth, resolver: externs.PopupRedirectResolver, bypassAuthState: boolean) => Promise<externs.UserCredential | null>;
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/model/popup_redirect.ts]: not sure i'm clear on how this works...?

File: packages-exp/auth-exp/src/core/user/additional_user_info.ts (hunk 6 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/user/additional_user_info.ts]: That shouldn't have been in there

File: packages-exp/auth-exp/src/core/user/reload.ts (hunk 7 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/user/reload.ts:57]: Done

File: packages-exp/auth-exp/src/core/auth/auth_impl.ts [hunk 10]
```diff
@@ -107,17 +108,21 @@ export class AuthImpl implements Auth, _FirebaseService {
...
         return;
       }
 
       this._isInitialized = true;
-      this.notifyAuthListeners();
     });
 
-    return this._initializationPromise;
+    // After initialization completes, throw any error caused by redirect flow
+    return this._initializationPromise.then(() => {
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/auth/auth_impl.ts:121]: why then instead of await?

File: packages-exp/auth-exp/src/core/auth/auth_impl.ts [hunk 11]
```diff
@@ -147,18 +152,40 @@ export class AuthImpl implements Auth, _FirebaseService {
...
 
-  private async initializeCurrentUser(): Promise<void> {
-    const storedUser = (await this.assertedPersistence.getCurrentUser()) as User | null;
+  private async initializeCurrentUser(
+    popupRedirectResolver?: externs.PopupRedirectResolver
+  ): Promise<void> {
+    // First check to see if we have a pending redirect event.
+    let storedUser = (await this.assertedPersistence.getCurrentUser()) as User | null;
+    if (popupRedirectResolver) {
+      await this.getOrInitRedirectPersistenceManager();
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/auth/auth_impl.ts:163]: should we have this method return the redirectUser?

File: packages-exp/auth-exp/src/core/providers/google.ts [hunk 12]
```diff
@@ -73,6 +73,7 @@ export class GoogleAuthProvider extends OAuthProvider {
 
   constructor() {
     super(externs.ProviderId.GOOGLE);
+    this.addScope('profile');
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/providers/google.ts:76]: add a test please

File: packages-exp/auth-exp/src/core/auth/auth_impl.ts [hunk 13]
```diff
@@ -180,6 +207,42 @@ export class AuthImpl implements Auth, _FirebaseService {
     return this.reloadAndSetCurrentUserOrClear(storedUser);
   }
 
+  private async tryRedirectSignIn(
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/auth/auth_impl.ts:210]: can we test this in auth_impl?

File: packages-exp/auth-exp/src/core/strategies/custom_token.ts [hunk 14]
```diff
@@ -43,7 +43,8 @@ export async function signInWithCustomToken(
   customToken: string
 ): Promise<externs.UserCredential> {
   const response: IdTokenResponse = await getIdTokenResponse(auth, {
-    token: customToken
+    token: customToken,
+    returnSecureToken: true
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/strategies/custom_token.ts:47]: this doesn't affect tests?

File: packages-exp/auth-exp/src/core/strategies/idp.ts [hunk 15]
```diff
@@ -85,20 +86,25 @@ class IdpCredential extends AuthCredential {
...
 
 /** @internal */
 export function _reauth(params: IdpTaskParams): Promise<UserCredential> {
   const { auth, user } = params;
   assert(user, AuthErrorCode.INTERNAL_ERROR, { appName: auth.name });
-  return _reauthenticate(user, new IdpCredential(params));
+  return _reauthenticate(
+    user,
+    new IdpCredential(params),
+    params.bypassAuthState
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/strategies/idp.ts:101]: add tests here if you can

File: packages-exp/auth-exp/src/core/user/additional_user_info.ts [hunk 16]
```diff
@@ -44,8 +44,8 @@ export function _fromIdTokenResponse(
     ];
     if (signInProvider) {
       const filteredProviderId =
-        providerId !== externs.ProviderId.ANONYMOUS &&
-        providerId !== externs.ProviderId.CUSTOM
+        signInProvider !== externs.ProviderId.ANONYMOUS &&
+        signInProvider !== externs.ProviderId.CUSTOM
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/user/additional_user_info.ts:48]: this wasn't caught by any tests?

File: packages-exp/auth-exp/src/core/user/reload.ts [hunk 17]
```diff
@@ -46,6 +46,19 @@ export async function _reloadWithoutSaving(user: User): Promise<void> {
...
+
+  const providerData = mergeProviderData(user.providerData, newProviderData);
+
+  // Preserves the non-nonymous status of the stored user, even if no more
+  // credentials (federated or email/password) are linked to the user. If
+  // the user was previously anonymous, then use provider data to update.
+  // On the other hand, if it was not anonymous before, it should never be
+  // considered anonymous now.
+  const oldIsAnonymous = user.isAnonymous;
+  const newIsAnonymous =
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/user/reload.ts:58]: do we have tests for this anonymous logic?

File: packages-exp/auth-exp/src/core/auth/auth_impl.ts (hunk 10 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/auth/auth_impl.ts:121]: If we await it, the error becomes part of the control flow of initialization. If there's an error with redirect during initialization, we still want auth to finish initializing

File: packages-exp/auth-exp/src/core/auth/auth_impl.ts (hunk 11 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/auth/auth_impl.ts:163]: no; the purpose of this method is to determine which user should be set as `currentUser`. The method first determines that, then updates it.

File: packages-exp/auth-exp/src/core/strategies/custom_token.ts (hunk 14 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/strategies/custom_token.ts:47]: It does now!

File: packages-exp/auth-exp/src/core/user/additional_user_info.ts (hunk 16 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/user/additional_user_info.ts:48]: No; the assumptions for anonymous and custom in the test were wrong and weren't caught in this logic branch. Updated.

File: packages-exp/auth-exp/src/core/user/reload.ts (hunk 17 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/user/reload.ts:58]: We do now!

File: packages-exp/auth-exp/src/core/auth/auth_impl.ts (hunk 10 above)
Comment (avolkovi) [reply packages-exp/auth-exp/src/core/auth/auth_impl.ts:123]: shouldn't we just wrap in a try catch then?

File: packages-exp/auth-exp/src/platform_browser/strategies/abstract_popup_redirect_operation.ts [hunk 18]
```diff
@@ -50,14 +50,16 @@ export abstract class AbstractPopupRedirectOperation
   private pendingPromise: PendingPromise | null = null;
   private eventManager: EventManager | null = null;
   readonly filter: AuthEventType[];
+  // private readonly bypassAuthState: boolean;
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/platform_browser/strategies/abstract_popup_redirect_operation.ts:53]: dead code

**Review by avolkovi — APPROVED** [review]

**Pull request merged**